arcaflow_plugin_horreum_client/horreum_client_plugin.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python3
+
+import sys
+import typing
+import requests
+import json
+
+from arcaflow_plugin_sdk import plugin
+from horreum_client_schema import InputParams, SuccessOutput, ErrorOutput
+
+
+@plugin.step(
+    id="horreum-client",
+    name="Horreum Client",
+    description="Uploads an object to the Horreum server",
+    outputs={"success": SuccessOutput, "error": ErrorOutput},
+)
+def horreum_client(
+    params: InputParams,
+) -> typing.Tuple[str, typing.Union[SuccessOutput, ErrorOutput]]:
+    auth_url = (
+        f"{params.horreum_keycloak_url}/realms/horreum/protocol/openid-connect/token"
+    )
+    auth_obj = {
+        "username": params.horreum_username,
+        "password": params.horreum_password,
+        "grant_type": "password",
+        "client_id": "horreum-ui",
+    }
+
+    try:
+        print("==> Authenticating with keycloak...")
+        auth_return = requests.post(auth_url, data=auth_obj, verify=params.tls_verify)
+        auth_return.raise_for_status()
+        token = json.loads(auth_return.text)["access_token"]
+
+    except (requests.ConnectionError, requests.HTTPError, requests.Timeout):
+        return "error", ErrorOutput(
+            f"Error communicating with server: {auth_return.text}"
+        )
+
+    send_url = (
+        f"{params.horreum_url}/api/run/data?"
+        f"test={params.test_name}"
+        f"&start={params.test_start_time}"
+        f"&stop={params.test_stop_time}"
+        f"&owner={params.test_owner}"
+        f"&access={params.test_access_rights}"
+    )
+    send_headers = {
+        "Authorization": f"Bearer {token}",
+        "content-type": "application/json",
+    }
+
+    try:
+        print("==> Uploading object to Horreum...")
+        send_return = requests.post(
+            send_url,
+            headers=send_headers,
+            json=params.data_object,
+            verify=params.tls_verify,
+        )
+
+    except (requests.ConnectionError, requests.HTTPError, requests.Timeout):
```

Comment:
Did you mean to have a `raise_for_status` here? Otherwise, I don't think you'll see the `HTTPError`. I see that you have a `send_return.status_code` check below, so this probably isn't important: you're clearly not depending on raising API errors.

Reply:
I added a `raise_for_status` to the first `requests.post` in the script, but I forgot to add it here. Fixed.

Reply:
Are you sure that you want to add a `raise_for_status()` here, given that you already handle errors a few lines later??

Reply:
Well, yeah: one or the other is good. There is the minor question of whether Horreum can ever return a 200 (or 100) status other than exactly `200` ... which in your explicit check is counted as a failure but with `raise_for_status()` would be treated as success. (E.g., in Pbench we detect a duplicate upload and return `200` ("OK") vs `201` ("CREATED"), and client code can detect the difference: but both are "success".)

Reply:
> There is the minor question of whether Horreum can ever return a 200 (or 100) status other than exactly 200 ... which in your explicit check is counted as a failure but with raise_for_status() would be treated as success.

That's actually a _new_ (and good) question:  with the addition of `raise_for_status()` another 2XX (or a 1XX or a 3XX) status won't be reported as an exception, and the error check at line 69 will report it (incorrectly, presumably) as an error.

Personally, I would use [`send_return.OK`](https://requests.readthedocs.io/en/latest/api/#requests.Response.ok) as the basis for checking for a failing HTTP status, and then perform a separate check for a Horreum-specific complaint (i.e., @dustinblack, what is the `not send_return.text.isdigit()` doing beyond what `send_return.OK` would do??).

Reply:
Thinking about this, I suspect for now we are most accurate in expecting exactly a 200 from Horreum, so the `raise_for_status` here is probably unnecessary.

The `not send_return.text.isdigit()` is explicitly checking that the text return we get from Horreum is an integer, which is the "good" return we should expect (this is the ID Horreum assigned to the uploaded run in a success case).